### ISSUE TITLE
#753 caused issues with tiled xfce4-terminal (xwayland)

### ISSUE THREAD

**Issue by jlindgren90** [labwc/labwc#785]
#753 caused a couple of issues that show up when tiling xfce4-terminal in xwayland mode:

1. When tiling the window (SnapToEdge), the window contents don't fully repaint
2. When untiling the window again, the natural geometry is not restored (the window keeps roughly the tiled geometry)

After some investigation, both issues stem from the fact that the tiled geometry does not exactly match the requested size increments of the terminal window, and the application is trying to adjust or "fix" the tiled geometry by sending a ConfigureRequest.

Before #753, we would honor the ConfigureRequest and set the requested size, resulting in a small gap between the terminal window and the edge of the screen. Now we ignore the ConfigureRequest but (in this case incorrectly) update the natural geometry.

The first issue needs to be fixed on the wlroots side. X11 window managers are free to ignore a client-initiated ConfigureRequest, but are supposed to send a synthetic ConfigureNotify with the unchanged window geometry in that case. wlroots doesn't currently send the synthetic event, leaving xfce4-terminal waiting (which is why it doesn't repaint). I'll submit a wlroots pull request shortly.

The second issue can be fixed on the labwc side, with some heuristics to decide whether or not to update the natural geometry from the ConfigureRequest. I'll submit a pull request for this too.

**Comment by jlindgren90** [comment]
Fix for first issue: https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4008

**Comment by Consolatis** [comment]
Regarding the second issue, as SnapToEdge (and SnapToRegion for that matter) are labwc internal states and not visible to the application, would it make sense to only use the change behavior introduced in #753 for maximized and fullscreen views?

**Comment by jlindgren90** [comment]
> would it make sense to only use the changed behavior introduced in #753 for maximized and fullscreen views?

I'm not sure, I saw the issue with a tiled view but I think it could also happen with a maximized or fullscreen view. My plan was basically to update the natural geometry only if it was empty/zero.

**Comment by Consolatis** [comment]
This might also have uncovered a third issue: I assume the terminal uses size hints so it seems we are not applying those for labwc initiated configure events (e.g. `view_move_resize()`). Otherwise xfce4-terminal wouldn't have to correct the sizes.

**Comment by jlindgren90** [comment]
> This might also have uncovered a third issue: I assume the terminal uses size hints so it seems we are not applying those for labwc initiated configure events (e.g. `view_move_resize()`). Otherwise xfce4-terminal wouldn't have to correct the sizes.

We apply the size hints for interactive resize (`process_cursor_resize()` calls `view_adjust_size()`) but not when maximizing/tiling. I think this is what we want: floating views should respect the size hints, but fullscreen/maximized/tiled views should fill the screen completely without leaving a gap. Openbox works the same way.

**Comment by Consolatis** [comment]
Right, missed that interactive resize applies them. So ignore that third issue.

In the meantime we could:
- only set natural_geometry when currently empty as you suggested, e.g. the desired usecase for #753
- in the case we go into that branch, send a synthetic configure ourselves so we don't have to depend on a future wlroots fix

**Comment by jlindgren90** [comment]
Draft PR for second issue: https://github.com/labwc/labwc/pull/786

**Comment by jlindgren90** [comment]
> send a synthetic configure ourselves so we don't have to depend on a future wlroots fix

Does wlroots expose the XCB connection to make this possible? If so, that could be a good option.

**Comment by Consolatis** [comment]
Sadly not. I stumbled across this while implementing the xdg <-> x11 bridge: even though wlr_xwayland_surface provides a pointer to `struct wlr_xwm` (which in turn has a `xcb_connection_t xcb_conn` member) we can't use it because its private.
Unless we draft an overlay struct and cast to that struct that is..

**Comment by jlindgren90** [comment]
> Edit: Couldn't we just use `wlr_xwayland_surface_configure()` though?

I tried that, didn't seem to work. I think that's relying on the X server to generate a "real" ConfigureNotify and it won't do that unless something actually changed.

**Comment by Consolatis** [comment]
Hmm. that sucks.
Just thinking out loud, alternative idea: We set the natural geometry but don't return early.
Then in the commit handler we check if the view is `!view_is_floating()` and then do an early out.
That would cause the state to be out of sync between the client and the xwayland surface vs `view->current` though which might cause further issues.

A proper solution would indeed be a new function in wlroots. But as we depend on 0.16 that doesn't help us at all.

**Comment by Consolatis** [comment]
This is some really nasty hack:
```c
struct xwm_overlay {
	void *xwayland;
	void *event_source;
	void *seat;
	uint32_t ping_timeout;
	xcb_atom_t atoms[68];
	xcb_connection_t *xcb_conn;
};
xcb_connection_t *xcb_conn = ((struct xwm_overlay *)server->xwayland->xwm)->xcb_conn;
```
Having the atoms before the xcb_connection is a bit unlucky. They have to be adjusted for 0.17, I think there were 2 additions since 0.16 (`NET_WM_STRUT_PARTIAL` and `WL_SURFACE_SERIAL`).

**Comment by jlindgren90** [comment]
I think for now I'd propose that we simply revert #753. Later (once the wlroots fix is merged and released) we can put a better solution in place for the original issue.

**Comment by Consolatis** [comment]
Regarding your upstream MR, does the new condition have to include the position as well? We might end up sending two `ConfigureNotify` events otherwise. I didn't look into the protocol details or other implementations regarding that, just want to make sure you are aware.

**Comment by jlindgren90** [comment]
I hope I understood the question correctly. If I did:

No, not based on my reading of the ICCCM spec. If the size changed, we can rely on the X server to send a "real" ConfigureNotify event and we don't need to synthesize one. If the position changed but not the size, we can't rely on the X server and need to synthesize a ConfigureNotify event.

The relevant section in ICCCM is:

> If the window manager decides to respond to a ConfigureRequest request by:
> 
> - Not changing the size, location, border width, or stacking order of the window at all.
> 
>     A client will receive a synthetic ConfigureNotify event that describes the (unchanged) geometry of the window. The (x,y) coordinates will be in the root coordinate system, adjusted for the border width the client requested, irrespective of any reparenting that has taken place. The border_width will be the border width the client requested. The client will not receive a real ConfigureNotify event because no change has actually taken place.
> 
> - Moving or restacking the window without resizing it or changing its border width.
> 
>     A client will receive a synthetic ConfigureNotify event following the change that describes the new geometry of the window. The event's (x,y) coordinates will be in the root coordinate system adjusted for the border width the client requested. The border_width will be the border width the client requested. The client may not receive a real ConfigureNotify event that describes this change because the window manager may have reparented the top-level window. If the client does receive a real event, the synthetic event will follow the real one.
> 
> - Resizing the window or changing its border width (regardless of whether the window was also moved or restacked).
> 
>     A client that has selected for StructureNotify events will receive a real ConfigureNotify event. Note that the coordinates in this event are relative to the parent, which may not be the root if the window has been reparented. The coordinates will reflect the actual border width of the window (which the window manager may have changed). The TranslateCoordinates request can be used to convert the coordinates if required.

**Comment by jlindgren90** [comment]
Closing this since the regressions should be fixed by the revert. #790 (draft) captures the remaining issue that we'll want to solve once the wlroots change is available.
